### PR TITLE
Hosttypyfy wpool

### PR DESCRIPTION
**Pull request by arcusfelis**
Proposed changes include:
* HostTypes in wpool
* Better specs in wpool

### CODE REVIEW

**Comment by mongoose-im** [comment]
[small_tests_21_3](https://circleci.com/gh/esl/MongooseIM/61073) / small_tests / 22c3d6e00dfc35a0dbd7cb88be32b938d9d036c4
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3170/61073/small_tests.21.3.8.17-1) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3170/61073/small_tests.21.3.8.17-1/small/ct_run.mongooseim@localhost.2021-07-06_13.22.46/lib.mongooseim.logs/run.2021-07-06_13.22.46/suite.log.html)

File: rebar.config [hunk 1]
```diff
@@ -71,7 +73,7 @@
   {cpool, "0.1.0"},
   {observer_cli, "1.5.4"},
   {nkpacket, {git, "https://github.com/michalwski/nkpacket.git", {ref, "f7c5349"}}},
-  {nksip, {git, "https://github.com/NetComposer/nksip.git", {ref, "1a29ef3"}}},
+  {nksip, {git, "https://github.com/arcusfelis/nksip.git", {branch, "mu-fix-types-1a29ef3"}}},
```
Comment (DenysGonchar) [inline rebar.config:76]: please submit PR with these fixes to the original repo.

File: src/cassandra/mongoose_cqerl_helper.erl [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+
+%% Violates cqerl:client() opaqueness
+-spec cqerl_client_to_pid(cqerl:client()) -> pid().
+cqerl_client_to_pid(Client) ->
+    {Pid, _Ref} = identity(Client),
+    Pid.
+
+%% Used to supress opaque check of dialyzer
+-spec identity(any()) -> any().
+identity(X) -> X.
```
Comment (DenysGonchar) [inline src/cassandra/mongoose_cqerl_helper.erl:12]: I don't really understand why we need it. please, illustrate it on some example?

File: src/jingle_sip/mod_jingle_sip.erl [hunk 3]
```diff
@@ -284,6 +271,28 @@ translate_to_sip(<<"session-terminate">>, Jingle, Acc) ->
             {error, item_not_found}
     end.
 
+%% There is some mess with types hapenning in nksip_uac:invite/3
```
Comment (DenysGonchar) [inline src/jingle_sip/mod_jingle_sip.erl:274]: didn't you fix it in your fork?

File: src/mam/mod_mam_utils.erl [hunk 4]
```diff
@@ -138,6 +138,8 @@ rsm_ns_binary() -> <<"http://jabber.org/protocol/rsm">>.
 -type archive_behaviour() :: mod_mam:archive_behaviour().
 -type archive_behaviour_bin() :: binary(). % `<<"roster">> | <<"always">> | <<"never">>'.
 
+%% calendar:rfc3339_string() type is not exported
+-type rfc3339_string() :: [byte(), ...].
```
Comment (DenysGonchar) [inline src/mam/mod_mam_utils.erl:142]: Should we put this type in some generic module (e.g. `mongooseim`)?

File: src/mod_stream_management.erl [hunk 5]
```diff
@@ -39,7 +39,11 @@
          remove_stale_smid_h/2
         ]).
 
--type smid() :: base64:ascii_binary().
+
+%% base64:ascii_binary() type is not exported
+-type ascii_binary() :: binary().
```
Comment (DenysGonchar) [inline src/mod_stream_management.erl:44]: the same question, should we put this type in some generic module (e.g. mongooseim)?

File: src/mongoose_redis.erl [hunk 6]
```diff
@@ -7,6 +7,7 @@
 -module(mongoose_redis).
 -author("bartlomiej.gorny@erlang-solutions.com").
 -include("mongoose.hrl").
+-include_lib("eredis/include/eredis.hrl").
```
Comment (DenysGonchar) [inline src/mongoose_redis.erl:10]: is `return_value()`  defined in `eredis.hrl`,  it seems to be a wrong approach

File: rebar.config [hunk 7]
```diff
@@ -178,7 +180,24 @@
    {override, jwerl, [{plugins, [rebar3_elixir, rebar3_hex]}]}
  ]}.
 
-{dialyzer, [{plt_extra_apps, [jid, cowboy, lasse, p1_utils, ranch, gen_fsm_compat, epgsql]}]}.
+{dialyzer, [
```
Comment (DenysGonchar) [inline rebar.config:183]: great job, but probably it's better to have it in a separate PR. it's just a wish for the future, no need to make a split of this one.

File: src/jlib.erl [hunk 8]
```diff
@@ -78,12 +78,16 @@
 -type rsm_in() :: #rsm_in{}.
 -type rsm_out() :: #rsm_out{}.
 
+%% calendar:rfc3339_string() type is not exported
+-type rfc3339_string() :: [byte(), ...].
```
Comment (DenysGonchar) [inline src/jlib.erl:82]: Should we put this type in some generic module (e.g. mongooseim)?

**Review by DenysGonchar — APPROVED** [review]
PR looks very good, I'd like to recheck just a few minor things before merging it.

**Comment by DenysGonchar** [comment]
Also that would be nice to run some integration tests with host-type specific pools.

File: src/jingle_sip/mod_jingle_sip.erl (hunk 3 above)
Comment (arcusfelis) [reply src/jingle_sip/mod_jingle_sip.erl:274]: nooo, there are still weird types in nksip.
 I've fixed one or two errors only.

File: src/cassandra/mongoose_cqerl_helper.erl (hunk 2 above)
Comment (arcusfelis) [reply src/cassandra/mongoose_cqerl_helper.erl:12]: without it you will get error that opaque type is used in mim code, while tuple is expected.

File: rebar.config (hunk 1 above)
Comment (arcusfelis) [reply rebar.config:76]: We use nksip from 5 years ago:

```
commit 1a29ef3e112ee0a8d8ac53bd6963ccc8f7879343
Author: Carlos Gonzalez <carlosj.gf@gmail.com>
Date:   Fri Nov 4 20:13:40 2016 +0100

    Fixed license
```

File: src/mongoose_redis.erl (hunk 6 above)
Comment (arcusfelis) [reply src/mongoose_redis.erl:10]: say it to eredis authors

**Comment by mongoose-im** [comment]
[small_tests_21_3](https://circleci.com/gh/esl/MongooseIM/61395) / small_tests / 91f57d2193a1c9f78496d479882849e7a6e7b9de
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3170/61395/small_tests.21.3.8.17-1) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3170/61395/small_tests.21.3.8.17-1/small/ct_run.mongooseim@localhost.2021-07-12_12.33.38/lib.mongooseim.logs/run.2021-07-12_12.33.38/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_22](https://circleci.com/gh/esl/MongooseIM/61655) / small_tests / 318756db689bd69278864aac428729c9f0314ce8
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3170/61655/small_tests.22.3.4.9-1) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3170/61655/small_tests.22.3.4.9-1/small/ct_run.mongooseim@localhost.2021-07-13_08.34.54/lib.mongooseim.logs/run.2021-07-13_08.34.54/suite.log.html)

File: src/event_pusher/mod_event_pusher_push.erl [hunk 9]
```diff
@@ -74,26 +74,31 @@
 %%--------------------------------------------------------------------
 %% gen_mod callbacks
 %%--------------------------------------------------------------------
--spec start(Host :: jid:server(), Opts :: list()) -> any().
+-spec start(HostType :: mongooseim:host_type(), Opts :: gen_mod:module_opts()) -> any().
```
Comment (chrzaszcz) [inline src/event_pusher/mod_event_pusher_push.erl:77]: Minor, I think that variable names should agree between the spec and the implementation. We could either change to `HostType` or leave `Host` in both places.

**Review by chrzaszcz — APPROVED** [review]
Nice, I like the new types, good work!

**Pull request merged**